### DISCUSSION TITLE
AttributeError when reflecting a Many-to-Many table

### DISCUSSION THREAD

**Discussion by AexisY** [sqlalchemy/sqlalchemy#6748] [Usage Questions] [answered]
Hi,

I am quite new to SQLAlchemy and want to use it in order to work on an existing dataBase using SQLAlchemy version 1.4.9.

I reflect the DB mith ORM stile:

```py
from sqlalchemy import create_engine,MetaData, Table, Column, Integer, String, ForeignKey, select
from sqlalchemy.engine import URL
from sqlalchemy.ext.automap import automap_base
from sqlalchemy.orm import relationship, sessionmaker, registry, Session


#connect_url comme from configuration

eng_tempus = create_engine(connect_url)

#Mapping en utilisant Auto Map, Voir https://docs.sqlalchemy.org/en/14/orm/extensions/automap.html
Base = automap_base()
Base.prepare(eng_tempus, reflect = True)

#Linking the classes to the tables
Users = Base.classes.Users
D_OF = Base.classes.D_OF
Event = Base.classes.Event
Machine = Base.classes.MACHINE
Code_evt = Base.classes.P_event
Status = Base.classes.Status
```

The last one "Status" table, is a Many-to-Many link between Event and Code_evt tables.
When I execute the last line an AttributError(key) is raised:
AttributeError: Status

The table is declaraed like this on the server (MS SQL2019):
```sql
CREATE TABLE "dbo"."Status"
(
   Seq int NOT NULL,
   Status int NOT NULL,
   CONSTRAINT PK_STATUS PRIMARY KEY (Seq,Status)
)
GO
ALTER TABLE "dbo"."Status"
ADD CONSTRAINT FK__Status__Seq__3F466844
FOREIGN KEY (Seq)
REFERENCES "dbo"."D_OF"(Sequence)
GO
ALTER TABLE "dbo"."Status"
ADD CONSTRAINT FK__Status__Status__403A8C7D
FOREIGN KEY (Status)
REFERENCES "dbo"."P_event"(ID)
GO
CREATE UNIQUE INDEX PK_STATUS ON "dbo"."Status"
(
  Seq,
  Status
)
GO
```

I read that SQLAlchemy is able to detect Many-to-Many relation when 2 foreign keys are used. But it doesn't work. Is there some think badly configured on my table? Or an option to be passed?

Thanks for the advice.

**Comment by CaselIT** [comment, marked as answer]
Hi,

You can use `Base.classes.keys()` to get a list of available classes.
Also the function [`classname_for_table`](https://docs.sqlalchemy.org/en/14/orm/extensions/automap.html#sqlalchemy.ext.automap.classname_for_table) can be used to find the name given to the class

**Reply by AexisY** [reply]
Thanks for the function .keys().
It can be very dinamic reflecting DB int this way!

**Reply by CaselIT** [reply]
@zzzeek The Property class (the type of Base.classes) does not seem to be documented. Maybe it should be?
Alternatively maybe can just say in the `Base.classes` documentation that's a mapping like object?

**Reply by zzzeek** [reply]
it would be good if Properties were linked and had a descriptive string at the class level, right now it has a one-liner.

**Reply by AexisY** [reply]
Ha CaseIIT,

With Base.classes.keys() the table 'Status' is not given.
Does it mean I can't maps this table? I will see furth more how request on many-to-many relationship work. May be it is a mishunderstanding.

Thanks for this tip.

**Reply by zzzeek** [reply]
see my answer.   Refer to D_OF.Statuses_collection and/or P_Event.Statuses_collection to refer to rows that make use of the "Status" table as intermediary table.

**Comment by zzzeek** [comment]
I dont think the many-to-many table is mapped as a class, it will be set as the "secondary" attribute on a relationship, https://docs.sqlalchemy.org/en/14/orm/extensions/automap.html#many-to-many-relationships describes this.   so you would want to look at the attributes on other classes that use  "Status" as an intermediary table.

you can map "Status" explicitly w/ automap also, so that you have it as a class: https://docs.sqlalchemy.org/en/14/orm/extensions/automap.html#specifying-classes-explicitly
